cmd/gateway-manta.go

```diff
@@ -0,0 +1,492 @@
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	} else {
+
+		var keyBytes []byte
+		if _, err = os.Stat(keyMaterial); err == nil {
+			keyBytes, err = ioutil.ReadFile(keyMaterial)
+			if err != nil {
+				return nil, fmt.Errorf("Error reading key material from %s: %s",
+					keyMaterial, err)
+			}
+			block, _ := pem.Decode(keyBytes)
+			if block == nil {
+				return nil, fmt.Errorf(
+					"Failed to read key material '%s': no key found", keyMaterial)
+			}
+
+			if block.Headers["Proc-Type"] == "4,ENCRYPTED" {
+				return nil, fmt.Errorf(
+					"Failed to read key '%s': password protected keys are\n"+
+						"not currently supported. Please decrypt the key prior to use.", keyMaterial)
+			}
+
+		} else {
+			keyBytes = []byte(keyMaterial)
+		}
+
+		signer, err = authentication.NewPrivateKeySigner(creds.SecretKey, keyBytes, creds.AccessKey)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	config := &triton.ClientConfig{
+		MantaURL:    endpoint,
+		AccountName: creds.AccessKey,
+		Signers:     []authentication.Signer{signer},
+	}
+	triton, err := storage.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+	triton.Client.HTTPClient = &http.Client{Transport: newCustomHTTPTransport()}
+
+	gateway := &tritonObjects{
+		client: triton,
+	}
+
+	return gateway, nil
+}
+
+// Shutdown - save any gateway metadata to disk
+// if necessary and reload upon next restart.
+func (t *tritonObjects) Shutdown() error {
+	return nil
+}
+
+// StorageInfo - Not relevant to Triton backend.
+func (t *tritonObjects) StorageInfo() (si StorageInfo) {
+	return si
+}
+
+//
+// ~~~ Buckets ~~~
+//
+
+// MakeBucketWithLocation - Create a new directory within manta.
+//
+// https://apidocs.joyent.com/manta/api.html#PutDirectory
+func (t *tritonObjects) MakeBucketWithLocation(bucket, location string) error {
+	ctx := context.Background()
+	err := t.client.Dir().Put(ctx, &storage.PutDirectoryInput{
+		DirectoryName: path.Join(mantaDefaultRootStore, bucket),
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetBucketInfo - Get directory metadata..
+//
+// https://apidocs.joyent.com/manta/api.html#GetObject
+func (t *tritonObjects) GetBucketInfo(bucket string) (bi BucketInfo, e error) {
+	var info BucketInfo
+	ctx := context.Background()
+	resp, err := t.client.Objects().Get(ctx, &storage.GetObjectInput{
+		ObjectPath: path.Join(mantaDefaultRootStore, bucket),
+	})
+	if err != nil {
+		return info, err
+	}
+
+	return BucketInfo{
+		Name:    bucket,
+		Created: resp.LastModified,
+	}, nil
+}
+
+// ListBuckets - Lists all Manta directories, uses Manta equivalent
+// ListDirectories.
+//
+// https://apidocs.joyent.com/manta/api.html#ListDirectory
+func (t *tritonObjects) ListBuckets() (buckets []BucketInfo, err error) {
+	ctx := context.Background()
+	dirs, err := t.client.Dir().List(ctx, &storage.ListDirectoryInput{
+		DirectoryName: path.Join(mantaDefaultRootStore),
+	})
+	if err != nil {
+		return nil, err
```

Comment:
all errors should be `errors.Trace()`ed and converted to object-errors ... take a look at `azureToObjectError()`